docs/hugo/content/architecture-decision-records/ADR-2022-02-Adoption-Policy.md

```diff
@@ -0,0 +1,127 @@
+# Adoption Policy 
+
+## Context
+
+Resources which exist in Azure prior to being created in Kubernetes are "adopted" by the operator when the resource
+is created in Kubernetes. Today this adoption process is automatic and implicit. As a user there isn't an easy way
+to tell if a resource you're about to `kubectl apply` already exists in Azure (will be adopted) or does not exist in
+Azure (will be created).
+
+We have recently proposed support for [reconcile-policy](https://github.com/Azure/azure-service-operator/pull/2060) which
+is related to adoption. Some of `reconcile-policy`'s primary use cases are for when the user does not want ASO to
+fully manage an adopted resource.
+
+The following are principles we should follow related to resource adoption:
+
+1. **The operator should adopt resources by default:** This is standard goal-seeking/desired-state behavior. 
+   If you create a resource in Kubernetes that says: "I expect there to be a Virtual Machine that looks like this
+   at location X" and that Virtual Machine already exists, then we are in the desired state.
+   The alternative would be to error if the resource already exists, which goes against the Kubernetes model.
+2. **Users must be able to instruct the operator to stop making changes to the backing Azure resource:**
+   This is enabled by `reconcile-policy`.
+
+   This is useful for the following scenarios:
+   - The operator is doing something wrong and the user needs an out.
+   - Pausing the operator temporarily while they make some manual fix via the Azure portal.
+   - Creating a resource in ASO purely to get ownership references to work. The resource in question is not managed
+     by ASO at all. It was created and managed some other way (portal, terraform, az-cli, etc).
+3. **Users must be able to instruct the operator to skip deletion of the backing resource:**
+   This is useful for scenarios such as moving ownership of resources from ASO in cluster A to ASO in cluster B.
+   This is enabled by `reconcile-policy`.
+
+### Open question: what should the default `reconcile-policy` be for resources which are adopted?
+
+There are two main options.
+
+#### Option 1: The default `reconcile-policy` for adopted resources is the same as for resources ASO created.
+
+ASO automatically adopts, updates, and will delete backing Azure resources as normal unless the user
+explicitly sets `reconcile-policy`.
+
+**Advantages:**
+- It's easy to explain what the behavior is. Simplistic users don't have to understand `reconcile-policy` (they won't
+  ever see it unless they set it themselves).
+
+**Disadvantages:**
+- Users have no good way to know their resource existed previously. They may `kubectl delete` it without realizing that
+  other things depend on this resource. Deletion (and some mutation) is destructive in Azure so lost data, etc cannot 
+  easily be recovered.
```

Comment:
I think this is a minor issue. Once I decide as a user to use ASO as the controller over my resources, I expect ASO to own the resources and don't care so much which already existed before and which didn't.
If I do care than I'll be explicit about it and set the `reconcile-policy` label, but then at least it's obvious in the source code of the resource (whether helm chart or bare yaml) that I for instance wanted to abandon a resource or skip the reconcile on it.

Reply:
I agree this is the case for users who understand the service operator model.

I have a concern about users who aren't experienced/knowledgable.

Consider what happens if a potential user starts auditioning ASO to see if it meets their needs and then inadvertently deletes something in Azure that they really shouldn't have? The most likely case would be a database or similar resource in their Dev/Test subscription. Worst case, they delete something from prod.

Is it really useful at that point for us to say "You should have understood what you were doing before you did it?"

While this is a minor issue overall, it wouldn't feel minor to that particular user.

I'm interested to see if there's a safe default that we can reasonably adopt, so that data-loss is less likely.

Reply:
Another important use case to consider is "update". For example, what happens if I create a LB with the name of a LB that already exists (without realizing it) but specify properties (eg. LB rules) that are different from the existing LB? Will ASO silently update my LB, potentially breaking other applications that rely on that resource? I can see how a "kubectl delete" is an explicit action but creating a resource with a name collision might be more surprising. 

A way to look at it another way is that performing a PUT in Azure on an existing resource does result in the resource being updated in the same pattern.

Reply:
> For example, what happens if I create a LB with the name of a LB that already exists (without realizing it) but specify properties (eg. LB rules) that are different from the existing LB?

Agree that this is definitely an important case to consider. There's nothing currently in this proposal that would automatically block mutation of an existing resource. In all cases that I could find, the mutations are reversible. I was thinking that something like shrinking PostgreSQL DB size would be non-reversible, but they disallow that operation (possibly because it's destructive). In your example, assuming you know what your old LB rules are, you can always put them back later, so at least there's an "out". The focus was on the delete case because there is a possibility that there's no recovery if you delete your database.

Option 2 outlined here would help with the delete case but does nothing for the update case. I don't see a way to satisfy this goal:
> **The operator should adopt resources by default:** This is standard goal-seeking/desired-state behavior. 
   If you create a resource in Kubernetes that says: "I expect there to be a Virtual Machine that looks like this
   at location X" and that Virtual Machine already exists, then we are in the desired state.
   The alternative would be to error if the resource already exists, which goes against the Kubernetes model.

and _also_ make adoption truly "safe". As you say, it's like submitting an ARM template for an existing resource. If you do it, it will modify the resource.

Reply:
I have amended this design to talk about two different questions here:
1. What is the default behavior?
2. Can users override that behavior?

My proposed choice is that the default behavior is to do what we're doing today (adopt resources if they already existed and manage them), _but_ unlike the original design (from nearly a year ago!), I also suggest we allow per-resource and operator-level overrides to this behavior so that if there are particular resources where users want a different behavior they can select that behavior.

Resolving this conversation as most of the points brought up in it have been incorporated into the design.